pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,472 @@
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(5).Info("checking", "podClaim", podClaim.Name)
+	templateName := podClaim.Source.ResourceClaimTemplateName
+	if templateName == nil {
+		return nil
+	}
+
+	claimName := resourceclaim.Name(pod, &podClaim)
+	claim, err := ec.claimLister.ResourceClaims(pod.Namespace).Get(claimName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if claim != nil {
+		if err := resourceclaim.IsForPod(pod, claim); err != nil {
+			return err
+		}
+		// Already created, nothing more to do.
+		logger.V(5).Info("claim already created", "podClaim", podClaim.Name, "resourceClaim", claimName)
+		return nil
+	}
+
+	template, err := ec.templateLister.ResourceClaimTemplates(pod.Namespace).Get(*templateName)
+	if err != nil {
+		return fmt.Errorf("resource claim template %q: %v", *templateName, err)
+	}
+
+	// Create the ResourceClaim with pod as owner.
+	isTrue := true
+	claim = &resourcev1alpha1.ResourceClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: claimName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "v1",
+					Kind:               "Pod",
+					Name:               pod.Name,
+					UID:                pod.UID,
+					Controller:         &isTrue,
+					BlockOwnerDeletion: &isTrue,
+				},
+			},
+			Annotations: template.Spec.ObjectMeta.Annotations,
+			Labels:      template.Spec.ObjectMeta.Labels,
+		},
+		Spec: template.Spec.Spec,
+	}
+	metrics.ResourceClaimCreateAttempts.Inc()
+	_, err = ec.kubeClient.ResourceV1alpha1().ResourceClaims(pod.Namespace).Create(ctx, claim, metav1.CreateOptions{})
+	if err != nil {
+		metrics.ResourceClaimCreateFailures.Inc()
+		return fmt.Errorf("create ResourceClaim %s: %v", claimName, err)
+	}
+	return nil
+}
+
+func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) error {
+	logger := klog.LoggerWithValues(klog.FromContext(ctx), "claim", klog.KRef(namespace, name))
+	ctx = klog.NewContext(ctx, logger)
+	claim, err := ec.claimLister.ResourceClaims(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(5).Info("nothing to do for claim, it is gone")
+			return nil
+		}
+		return err
+	}
+
+	// Check if the ReservedFor entries are all still valid.
+	valid := make([]resourcev1alpha1.ResourceClaimConsumerReference, 0, len(claim.Status.ReservedFor))
+	for _, reservedFor := range claim.Status.ReservedFor {
+		if reservedFor.APIGroup == "" &&
+			reservedFor.Resource == "pods" {
+			// A pod falls into one of three categories:
+			// - we have it in our cache -> don't remove it until we are told that it got removed
+			// - we don't have it in our cache anymore, but we have seen it before -> it was deleted, remove it
+			// - not in our cache, not seen -> double-check with API server before removal
+
+			keepEntry := true
+			if ec.deletedObjects.Has(reservedFor.UID) {
+				// We know that the pod was deleted. This is
+				// easy to check and thus is done first.
+				keepEntry = false
```

Comment:
what is this optimizing for?

If we got an event for the pod deletion, it's already not in the lister.

Reply:
I'll add a comment before the if check:
> Tracking deleted pods in the LRU cache is an optimization. Without this
> cache, the code would have to do the API call below for every deleted
> pod to ensure that the pod really doesn't exist. With the LRU cache, most
> of the time the pod will be recorded as deleted and the API call can be
> avoided.